src/LibraryManager/Providers/BaseProvider.cs

```diff
@@ -0,0 +1,297 @@
+            }
+
+            return null;
+        }
+
+        private bool IsLibraryUpToDate(ILibraryInstallationState state)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(state.Name) && !string.IsNullOrEmpty(state.Version))
+                {
+                    string cacheDir = Path.Combine(CacheFolder, state.Name, state.Version);
+                    string destinationDir = Path.Combine(HostInteraction.WorkingDirectory, state.DestinationPath);
+
+                    foreach (string sourceFile in state.Files)
+                    {
+                        var destinationFile = new FileInfo(Path.Combine(destinationDir, sourceFile).Replace('\\', '/'));
+                        var cacheFile = new FileInfo(Path.Combine(cacheDir, sourceFile).Replace('\\', '/'));
+
+                        if (!destinationFile.Exists || !cacheFile.Exists || !FileHelpers.AreFilesUpToDate(destinationFile, cacheFile))
+                        {
+                            return false;
+                        }
+                    }
+                }
+            }
+            catch
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Copies ILibraryInstallationState files to cache
+        /// </summary>
+        /// <param name="state"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async Task<ILibraryOperationResult> RefreshCacheAsync(ILibraryInstallationState state, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return LibraryOperationResult.FromCancelled(state);
+            }
+
+            string libraryDir = Path.Combine(CacheFolder, state.Name, state.Version);
+
+            try
+            {
+                var librariesMetadata = new List<CacheFileMetadata>();
+                foreach (string sourceFile in state.Files)
+                {
+                    string cacheFile = Path.Combine(libraryDir, sourceFile);
+                    string url = GetDownloadUrl(state, sourceFile);
+
+                    var newEntry = new CacheFileMetadata(url, cacheFile);
+                    if (!librariesMetadata.Contains(newEntry))
+                    {
+                        librariesMetadata.Add(new CacheFileMetadata(url, cacheFile));
+                    }
+                }
+                await _cacheService.RefreshCacheAsync(librariesMetadata, cancellationToken);
+            }
+            catch (ResourceDownloadException ex)
+            {
+                HostInteraction.Logger.Log(ex.ToString(), LogLevel.Error);
+                return new LibraryOperationResult(state, PredefinedErrors.FailedToDownloadResource(ex.Url));
+            }
+            catch (OperationCanceledException)
+            {
+                return LibraryOperationResult.FromCancelled(state);
```

Comment:
We just swallow these? Do they happen a lot during normal execution? Is it basically a timeout? If it means we didn't refresh successfully, should we log this anyway?

Reply:
This is the type thrown when a task is cancelled - so if we have cancel-able operations (like background tasks in VS), this will catch it here.

I don't know if they happen often, since we don't test cancellation heavily.

Reply:
No, these would happen if the user cancels the operation, not when the operation times out.  And we do log it (see `LogMessageGenerator.GetOperationSummaryString()`).